docs/man/man3/Receiver.3
.TH "Receiver" 3 "Version v01.01r" "libcnet" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Receiver
.SH SYNOPSIS
.br
.PP
.PP
\fR#include <receiver\&.h>\fP
.SS "Data Fields"

.in +1c
.ti -1c
.RI "struct protoent * \fB_proto\fP"
.br
.RI "The protocol used\&. "
.ti -1c
.RI "struct sockaddr_in \fB_address\fP"
.br
.RI "The IP address on which we are binded\&. "
.ti -1c
.RI "int \fB_socket\fP"
.br
.RI "File descriptor for the created socket\&. "
.ti -1c
.RI "bool \fB_running\fP"
.br
.RI "If the receiver in receiving or not\&. "
.ti -1c
.RI "bool \fB_verbose\fP"
.br
.RI "Enable verbosity\&. "
.ti -1c
.RI "struct \fBTimer\fP * \fB_timer\fP"
.br
.RI "An optional \fBTimer\fP\&. "
.ti -1c
.RI "sem_t * \fB_mutex\fP"
.br
.RI "A semaphore to synchronize with the \fBSender\fP\&. "
.ti -1c
.RI "\fBMessageQueue\fP * \fB_queue\fP"
.br
.RI "The message queue, essentially a linked list\&. "
.in -1c
.SH "Detailed Description"
.PP 
Struct representing a \fBReceiver\fP\&. Essentially, it is an always running thread listening for incoming packets on a defined protocol on a given port\&.
.PP
It has the following fields:
.IP "\(bu" 2
\fR_proto\fP is a pointer to a \fRprotoent\fP structure defining which protocol is used
.IP "\(bu" 2
\fR_address\fP is a \fRsockaddr_in\fP value containing the IP address on which the receiver is binded
.IP "\(bu" 2
\fR_socket\fP is just the file descriptor of the created socket
.IP "\(bu" 2
\fR_running\fP a boolean indicating if the receiver is running or not
.IP "\(bu" 2
\fR_verbose\fP A flag to enable verbosity
.IP "\(bu" 2
\fR_timer\fP A timer shared with the \fBSender\fP
.IP "\(bu" 2
\fR_mutex\fP A simple Mutex to syncrhonize with the sender 
.PP

.PP
Definition at line \fB63\fP of file \fBreceiver\&.h\fP\&.
.SH "Field Documentation"
.PP 
.SS "struct sockaddr_in _address"

.PP
The IP address on which we are binded\&. 
.PP
Definition at line \fB67\fP of file \fBreceiver\&.h\fP\&.
.SS "sem_t* _mutex"

.PP
A semaphore to synchronize with the \fBSender\fP\&. 
.PP
Definition at line \fB72\fP of file \fBreceiver\&.h\fP\&.
.SS "struct protoent* _proto"

.PP
The protocol used\&. 
.PP
Definition at line \fB66\fP of file \fBreceiver\&.h\fP\&.
.SS "\fBMessageQueue\fP* _queue"

.PP
The message queue, essentially a linked list\&. 
.PP
Definition at line \fB73\fP of file \fBreceiver\&.h\fP\&.
.SS "bool _running"

.PP
If the receiver in receiving or not\&. 
.PP
Definition at line \fB69\fP of file \fBreceiver\&.h\fP\&.
.SS "int _socket"

.PP
File descriptor for the created socket\&. 
.PP
Definition at line \fB68\fP of file \fBreceiver\&.h\fP\&.
.SS "struct \fBTimer\fP* _timer"

.PP
An optional \fBTimer\fP\&. 
.PP
Definition at line \fB71\fP of file \fBreceiver\&.h\fP\&.
.SS "bool _verbose"

.PP
Enable verbosity\&. 
.PP
Definition at line \fB70\fP of file \fBreceiver\&.h\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for libcnet from the source code\&.


CHANGELOG.md
# CHANGELOG

## libcnet v01.01r -- 29/06/2024 - 04/07/2024

##### Addition

- Added TCP Header and Packets structure and functions to `ip.h` and `ip.c`
- Added functions for IP Packets to include TCP Packets into `ip.h` and `ip.c`
- Added a version header files containing the current version of the library `utils/version.h`
- Added documentation using Doxygen. Modified all docstrings in according to Doxygen documentation
- Added crafting functions to the Sender for sending TCP Packets, `Sender_craftTcp` and `Sender_fillTcpHeader`
- Added `PseudoHeader` struct to `ip.h` and functions to compute Checksum for ICMP, UDP and TCP

##### Small Changes

- Changed `snprintf` into `sprintf` into `ControlBits_toBin` function of `ip.c`
- Added TCP Option printing informations to the `TcpHeader_printInfo` function

##### Fixed Bugs

- When wrapping a packet into an IP Packet it causes the address of the payload being copied.
  This can cause a double free error when trying to call the `IpPacket_delete` function. To
  avoid this behaviour, first the Header and then the payload are copied into new memory
  locations. Before it was `memcpy(dstpacket, srcpacket, size)` now is
  `memcpy(dstpacket->hdr, srcpacket->hdr, hdr_size)` and `memcpy(dstpacket->payload, srcpacket->payload, psize)`

- When computing the checksum it was jumping two bytes, now it does not. Notice that it computes the checksum
  correctly both for ICMP and UDP Packets, but it does not for TCP Packets and complete IP Packets.

- Correct the bug when creating the TCP Option and then encoding flipping the bytes in an incorrect way.

src/receiver.c
#include "receiver.h"

Receiver* Receiver_new(
    const char* _interface, const u_int16_t _port, const char* _proto, const bool _verbose
) {
    struct sockaddr_in addr;
    int socketfd;
    struct protoent *proto = getprotobyname(_proto);

    char _addr[INET_ADDRSTRLEN];
    getInterfaceIp(_interface, _addr);

    socketfd = socket(AF_INET, SOCK_RAW, proto->p_proto);
    if (socketfd == -1) handle_error("socket", socketfd);

    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(_port);
    inet_pton(AF_INET, _addr, &addr.sin_addr);

    if (bind(socketfd, (const struct sockaddr*)&addr, sizeof(addr)) < 0) 
    {
        handle_error("bind", socketfd);
    }

    if (_verbose)
    {
        printf("[*] %s Receiver binded on (%s, %hu)\n", proto->p_name, _addr, _port);
    }

    Receiver* recv = (Receiver*)malloc(sizeof(Receiver));
    recv->_address = addr;
    recv->_proto = proto;
    recv->_socket = socketfd;
    recv->_running = false;
    recv->_verbose = _verbose;
    recv->_timer = NULL;
    recv->_mutex = NULL;
    recv->_queue = MessageQueue_new(100);

    return recv;
}

void Receiver_delete(Receiver* _self)
{
    shutdown(_self->_socket, 2);
    MessageQueue_delete(_self->_queue);
    free(_self);
}

void *Receiver_run(void* _self)
{
    u_int32_t addrn = ((Receiver*)_self)->_address.sin_addr.s_addr;
    u_int16_t port = ((Receiver*)_self)->_address.sin_port;
    char* pname = ((Receiver*)_self)->_proto->p_name;
    
    char addr[INET_ADDRSTRLEN];
    addressNumberToString(addrn, addr, true);
    
    if (((Receiver*)_self)->_verbose)
    {
        printf("[*] %s Receiver starting on (%s, %hu)\n", pname, addr, ntohs(port));
    }

    size_t ip_size = (IP_PAYLOAD_MAX_SIZE + IP_HEADER_SIZE);
    socklen_t socklen = sizeof(((Receiver*) _self)->_address);
    char buff[ip_size];

    struct Response *resp = (struct Response*)malloc(sizeof(struct Response));
    resp->_buffer = (char *)malloc(ip_size * sizeof(char));
    double rtt;

    while (((Receiver*)_self)->_running)
    {

        if (((Receiver*)_self)->_mutex != NULL)
        {
            semaphore_wait(((Receiver*)_self)->_mutex, "Receiver_run");
        }

        // Receive a message from the socket
        ssize_t retval = recvfrom(
            ((Receiver*) _self)->_socket, buff, ip_size, MSG_DONTWAIT,
            (struct sockaddr*)&((Receiver*) _self)->_address, &socklen
        );
        
        if (retval < 0)
        {
            if (((Receiver*)_self)->_mutex != NULL)
            {
                semaphore_post(((Receiver*)_self)->_mutex, "Receiver_run");
            }

            continue;
        }

        if (((Receiver*)_self)->_verbose) 
        {
            printf("[*] %s Receiver received %ld bytes\n", pname, retval);
        }

        rtt = ((Receiver*)_self)->_timer != NULL ? Timer_getDeltaTime(((Receiver*)_self)->_timer) : 0.0;

        // Create the response
        memcpy(resp->_buffer, buff, retval);
        resp->_rtt = rtt;
        resp->_size = retval;

        // Push the response in the message queue
        MessageQueue_push(((Receiver*)_self)->_queue, resp, sizeof(struct Response));

        if (((Receiver*)_self)->_mutex != NULL)
        {
            semaphore_post(((Receiver*)_self)->_mutex, "Receiver_run");
        }
    }

    if (((Receiver*)_self)->_timer != NULL)
    {
        Timer_stop(((Receiver*)_self)->_timer);
    }

    Response_free(resp);

    if (((Receiver*)_self)->_verbose)  printf("[*] Receiver stopped\n");
    return NULL;
}

void Receiver_start(Receiver* _self)
{
    _self->_running = true;

    // Start the timer if is not NULL
    if (_self->_timer != NULL) Timer_start(_self->_timer);

    pthread_t recv_thread;
    pthread_create(&recv_thread, NULL, Receiver_run, _self);
    pthread_detach(recv_thread);
}

void Receiver_stop(Receiver* _self)
{
    _self->_running = false;
    sleep(1.0);
}

void Receiver_setTimer(Receiver* _self, struct Timer* _timer)
{
    _self->_timer = _timer;
}

void Receiver_setSemaphore(Receiver* _self, sem_t* _mutex)
{
    _self->_mutex = _mutex;
}

void Response_free(struct Response* _self)
{
    free(_self->_buffer);
    free(_self);
}